recursion/core/src/memory/mod.rs

```diff
@@ -2,34 +2,73 @@ mod air;
 mod columns;
 
 use crate::air::Block;
-use sp1_derive::AlignedBorrow;
+pub use columns::*;
 
+#[allow(clippy::manual_non_exhaustive)]
 #[derive(Debug, Clone)]
 pub struct MemoryRecord<F> {
     pub addr: F,
     pub value: Block<F>,
-    pub timestamp: F,
     pub prev_value: Block<F>,
+    pub timestamp: F,
     pub prev_timestamp: F,
 }
 
-#[derive(AlignedBorrow, Default, Debug, Clone)]
-#[repr(C)]
-pub struct MemoryReadWriteCols<T> {
-    pub addr: T,
-    pub value: Block<T>,
-    pub timestamp: T,
-    pub prev_value: Block<T>,
-    pub prev_timestamp: T,
+impl<F: Clone> MemoryRecord<F> {
+    pub fn new_write(
+        addr: F,
+        value: Block<F>,
+        timestamp: F,
+        prev_value: Block<F>,
+        prev_timestamp: F,
+    ) -> Self {
+        Self {
+            addr,
+            value,
+            prev_value,
+            timestamp,
+            prev_timestamp,
+        }
+    }
+
+    pub fn new_read(addr: F, value: Block<F>, timestamp: F, prev_timestamp: F) -> Self {
+        Self {
+            addr,
+            value: value.clone(),
+            prev_value: value,
+            timestamp,
+            prev_timestamp,
+        }
+    }
 }
 
 impl<T: Clone> MemoryReadWriteCols<T> {
     pub fn populate(&mut self, record: &MemoryRecord<T>) {
-        self.addr = record.addr.clone();
-        self.value = record.value.clone();
-        self.timestamp = record.timestamp.clone();
+        self.access.prev_timestamp = record.prev_timestamp.clone();
+        self.access.value = record.value.clone();
         self.prev_value = record.prev_value.clone();
-        self.prev_timestamp = record.prev_timestamp.clone();
+    }
+}
+
+impl<T: Clone> MemoryReadCols<T> {
+    pub fn populate(&mut self, record: &MemoryRecord<T>) {
+        self.access.prev_timestamp = record.prev_timestamp.clone();
+        self.access.value = record.value.clone();
+    }
+}
+
+impl<T: Clone> MemoryReadWriteSingleCols<T> {
+    pub fn populate(&mut self, record: &MemoryRecord<T>) {
+        self.access.prev_timestamp = record.prev_timestamp.clone();
+        self.access.value = record.value.0[0].clone();
```

Comment:
This is somewhat difficult because `T` isn't necessarily a field element here, and it's usually not - it's `impl Into<E>`, for some type `E: AbstractField`. However, I don't know how to write that here without making `MemoryReadWriteSingleCols` generic over both `T` and `E`, which feels quite unnecessary.